include/fastrtps/utils/shared_mutex.hpp

```diff
@@ -142,17 +171,144 @@ class shared_mutex
                 gate2_.notify_one();
             }
         }
-        else
+        else if (num_readers == n_readers_ - 1)
         {
-            if (num_readers == n_readers_ - 1)
-            {
-                gate1_.notify_one();
-            }
+            gate1_.notify_one();
+        }
+    }
+
+};
+
+// implementation not locking readers on behalf of writers
+
+template<>
+class shared_mutex<shared_mutex_type::PTHREAD_RWLOCK_PREFER_READER_NP>
+    : public shared_mutex_base
+{
+    count_t writer_waiting_ = 0;
+
+public:
+
+    void lock()
+    {
+        std::unique_lock<mutex_t> lk(mut_);
+        ++writer_waiting_;
+        while (state_ & n_readers_ || state_ & write_entered_)
+        {
+            gate1_.wait(lk);
+        }
+        state_ |= write_entered_;
+        --writer_waiting_;
+    }
+
+    void unlock_shared()
+    {
+        std::lock_guard<mutex_t> _(mut_);
+        count_t num_readers = (state_ & n_readers_) - 1;
+        state_ &= ~n_readers_;
+        state_ |= num_readers;
+
+        if ((writer_waiting_ && num_readers == 0)
+                || (num_readers == n_readers_ - 1))
+        {
+            gate1_.notify_one();
+        }
+    }
+
+};
+
+// Debugger wrapper class that provides insight
+template<class sm>
+class debug_wrapper : public sm
```

Comment:
There is an inconsistency in this class.

Sometimes (`lock()`, `shared_lock()`) the base class is called, and then mutex `wm_` is taken to modify the debug information. Other times (`try_lock()`, `unlock()`, `try_lock_shared()`, and `unlock_shared()`), mutex `wm_` is taken before calling the base class.

I think we should always do it in the same order.

Reply:
Ok, but note that an ABBA is impossible because the base class ignores the outer mutex existence and its methods do not call external code.